src/modules/vision_target_estimator/Position/KF_xyzb_v_decoupled_moving.cpp

```diff
@@ -0,0 +1,149 @@
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file KF_xyzb_v_decoupled_moving.cpp
+ * @brief Filter to estimate the pose of moving targets. State: [r, vd, b, at, vt]
+ *
+ * @author Jonas Perolini <jonspero@me.com>
+ *
+ */
+
+#include "KF_xyzb_v_decoupled_moving.h"
+#include "python_derivation/generated/decoupled_moving_xyzb_v/predictCov.h"
+#include "python_derivation/generated/decoupled_moving_xyzb_v/computeInnovCov.h"
+
+namespace vision_target_estimator
+{
+
+void KF_xyzb_v_decoupled_moving::predictState(float dt, float acc)
+{
+	// _state [r, vd, b, at, vt]
+	// idx    [0,   1,  2,   3,   4 ]
+
+	const float tmp0 = 0.5f * dt * dt;
+
+	_state(0, 0) = _state(0, 0) - tmp0 * acc + tmp0 * _state(3, 0) - dt * _state(1, 0) + dt * _state(4, 0);
+	_state(1, 0) = _state(1, 0) + acc * dt;
+	_state(4, 0) = _state(4, 0) + dt * _state(3, 0);
+}
+
+void KF_xyzb_v_decoupled_moving::predictCov(float dt)
+{
+	matrix::Matrix<float, 5, 5> cov_updated;
+	sym::Predictcov(dt, _input_var, _bias_var, _acc_var, _covariance, &cov_updated);
+	_covariance = cov_updated;
+}
+
+
+bool KF_xyzb_v_decoupled_moving::update()
+{
+	// Avoid zero-division
+	if (fabsf(_innov_cov) < 1e-6f) {
+		return false;
+	}
+
+	const float beta = _innov / _innov_cov * _innov;
+
+	// Normalized innovation Squared threshold. Checks whether innovation is consistent with innovation covariance.
+	if (beta > _nis_threshold) {
+		return false;
+	}
+
+	const matrix::Matrix<float, 5, 1> kalmanGain = _covariance * _meas_matrix.transpose() / _innov_cov;
+
+	_state = _state + kalmanGain * _innov;
+
+	_covariance = _covariance - kalmanGain * _meas_matrix * _covariance;
+
+	return true;
+}
+
+void KF_xyzb_v_decoupled_moving::setH(matrix::Vector<float, 15> h_meas, int direction)
+{
+	// _state [rx, ry, rz, vdx, vdy, vdz, bx, by, bz, atx, aty, atz, vtx, vty, vtz]
+	// idx    [0,   1,  2,   3,   4,   5,  6,  7,  8,   9,  10,  11,  12,  13,  14]
```

Comment:
enums are made for that.

Reply:
Done in https://github.com/PX4/PX4-Autopilot/pull/22008/commits/ad989227c1e7608835019cd8009e41d85174702b

Reply:
Improved in https://github.com/PX4/PX4-Autopilot/pull/22008/commits/53b45be8e28c1f35870dbc7e2a0f8ab3f509d0ac